lib/common/tests/rules/pcmk__evaluate_attr_expression_test.c

```diff
@@ -0,0 +1,718 @@
+        PCMK_XA_OPERATION "='" PCMK_VALUE_EQ "' "       \
+        PCMK_XA_ATTRIBUTE "='foo' "                     \
+        PCMK_XA_VALUE "='bar' />"
+
+static void
+type_string_passes(void **state)
+{
+    assert_attr_expression(EXPR_TYPE_STRING_PASSES, pcmk_rc_ok);
+}
+
+#define EXPR_TYPE_STRING_FAILS                          \
+        "<" PCMK_XE_EXPRESSION " " PCMK_XA_ID "='e' "   \
+        PCMK_XA_TYPE "='" PCMK_VALUE_STRING "' "        \
+        PCMK_XA_OPERATION "='" PCMK_VALUE_EQ "' "       \
+        PCMK_XA_ATTRIBUTE "='foo' "                     \
+        PCMK_XA_VALUE "='bat' />"
+
+static void
+type_string_fails(void **state)
+{
+    assert_attr_expression(EXPR_TYPE_STRING_FAILS, pcmk_rc_op_unsatisfied);
+}
+
+#define EXPR_TYPE_INTEGER_PASSES                        \
+        "<" PCMK_XE_EXPRESSION " " PCMK_XA_ID "='e' "   \
+        PCMK_XA_TYPE "='" PCMK_VALUE_INTEGER "' "       \
+        PCMK_XA_OPERATION "='" PCMK_VALUE_EQ "' "       \
+        PCMK_XA_ATTRIBUTE "='num' "                     \
+        PCMK_XA_VALUE "='10' />"
+
+static void
+type_integer_passes(void **state)
+{
+    assert_attr_expression(EXPR_TYPE_INTEGER_PASSES, pcmk_rc_ok);
+}
+
+#define EXPR_TYPE_INTEGER_FAILS                         \
+        "<" PCMK_XE_EXPRESSION " " PCMK_XA_ID "='e' "   \
+        PCMK_XA_TYPE "='" PCMK_VALUE_INTEGER "' "       \
+        PCMK_XA_OPERATION "='" PCMK_VALUE_EQ "' "       \
+        PCMK_XA_ATTRIBUTE "='num' "                     \
+        PCMK_XA_VALUE "='11' />"
+
+static void
+type_integer_fails(void **state)
+{
+    assert_attr_expression(EXPR_TYPE_INTEGER_FAILS, pcmk_rc_op_unsatisfied);
+}
+
+#define EXPR_TYPE_NUMBER_PASSES                         \
+        "<" PCMK_XE_EXPRESSION " " PCMK_XA_ID "='e' "   \
+        PCMK_XA_TYPE "='" PCMK_VALUE_NUMBER "' "        \
+        PCMK_XA_OPERATION "='" PCMK_VALUE_EQ "' "       \
+        PCMK_XA_ATTRIBUTE "='num' "                     \
+        PCMK_XA_VALUE "='10.0' />"
+
+static void
+type_number_passes(void **state)
+{
+    assert_attr_expression(EXPR_TYPE_NUMBER_PASSES, pcmk_rc_ok);
+}
+
+#define EXPR_TYPE_NUMBER_FAILS                          \
+        "<" PCMK_XE_EXPRESSION " " PCMK_XA_ID "='e' "   \
+        PCMK_XA_TYPE "='" PCMK_VALUE_NUMBER "' "        \
+        PCMK_XA_OPERATION "='" PCMK_VALUE_EQ "' "       \
+        PCMK_XA_ATTRIBUTE "='num' "                     \
+        PCMK_XA_VALUE "='10.1' />"
+
+static void
+type_number_fails(void **state)
+{
+    assert_attr_expression(EXPR_TYPE_NUMBER_FAILS, pcmk_rc_op_unsatisfied);
+}
+
+#define EXPR_TYPE_VERSION_PASSES                        \
+        "<" PCMK_XE_EXPRESSION " " PCMK_XA_ID "='e' "   \
+        PCMK_XA_TYPE "='" PCMK_VALUE_VERSION "' "       \
+        PCMK_XA_OPERATION "='" PCMK_VALUE_GT "' "       \
+        PCMK_XA_ATTRIBUTE "='ver' "                     \
+        PCMK_XA_VALUE "='3.4.9' />"
+
+static void
+type_version_passes(void **state)
```

Comment:
We may want a passing equality test somewhere for version, possibly with `3.5` vs `3.5.0`

Reply:
might as well, though that borders on unit-testing `compare_version()`

Reply:
agreed, not tied to it